pkg/event/target/mqtt.go

```diff
@@ -179,54 +178,30 @@ func NewMQTTTarget(id string, args MQTTArgs, doneCh <-chan struct{}, loggerOnce
 	// Connect() should be successful atleast once to publish events.
 	token := client.Connect()
 
+	if token.Wait() && token.Error() != nil {
+		return nil, token.Error()
+	}
+
 	target := &MQTTTarget{
 		id:         event.TargetID{ID: id, Name: "mqtt"},
 		args:       args,
 		client:     client,
 		loggerOnce: loggerOnce,
 	}
 
-	// Retries until the clientID gets registered.
-	retryRegister := func() {
```

Comment:
We need the retry register because the mqtt client should register its clientID and obtain a token atleast once.. Publishing events will not work if the client publishing it hasn't yet obtained a token (got its clientID registered) from the mqtt server/broker.

This logic was added since the mqtt library did not retry the token registration process.

This will not work if we start MinIO server with an inactive mqtt broker and queueDir enabled in the config. It will not replay the events even when the broker comes back online. Because we never tried to re-connect the existing client and obtain a token. I once raised a PR - https://github.com/eclipse/paho.mqtt.golang/pull/281 on this.

Reply:
To reproduce this :- 

```
package main

import (
	mqtt "github.com/eclipse/paho.mqtt.golang"
	"fmt"
	"time"
)


func main() {
	options := mqtt.NewClientOptions().
		SetClientID("").
		SetCleanSession(true). 
		AddBroker("tcp://localhost:1883")

	client := mqtt.NewClient(options)

	token := client.Connect()

	go func() {
		for {
			// Message will not be sent until the client obtains a token.
			token = client.Publish("minio123", 2, false, "hello")
			if token.Error() != nil {
				fmt.Println("message not sent. Token error: ", token.Error().Error())
			} else {
				fmt.Println("message sent")
				break
			}
			time.Sleep(2* time.Second)
		}
	}()

	for {}
}
```

- Run this with an inactive broker
- Start the broker after some time
- The events will keep failing even after that since the token for that client is not yet obtained

Reply:
The flow has changed @Praveenrajmani PTAL